### ISSUE TITLE
gci does not take effect when used with `...` style

### ISSUE THREAD

**Issue by howardjohn** [golangci/golangci-lint#1351]
Running command on `./...` style reference, no issues found
```console
build-tools:/work$ golangci-lint run -v -c ./common/config/.golangci-format.yml ./tests/integration/telemetry/tracing/...
INFO [config_reader] Used config file common/config/.golangci-format.yml
INFO [lintersdb] Active 2 linters: [gci goimports]
INFO [loader] Go packages loading at mode 575 (types_sizes|deps|imports|name|compiled_files|exports_file|files) took 1.297977365s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 286.907µs
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner] processing took 4.451µs with stages: max_same_issues: 718ns, nolint: 614ns, skip_dirs: 465ns, max_from_linter: 372ns, path_prettifier: 305ns, cgo: 283ns, filename_unadjuster: 256ns, source_code: 217ns, uniq_by_line: 213ns, autogenerated_exclude: 207ns, skip_files: 125ns, path_shortener: 124ns, identifier_marker: 122ns, diff: 121ns, exclude: 66ns, max_per_file_from_linter: 63ns, exclude-rules: 51ns, severity-rules: 45ns, sort_results: 42ns, path_prefixer: 42ns
INFO [runner] linters took 296.246672ms with stages: goanalysis_metalinter: 296.202546ms
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 17 samples, avg is 83.9MB, max is 139.3MB
INFO Execution took 1.598983266s
```

Running command on single file, issues are (correctly) found:
```console
build-tools:/work$ golangci-lint run -v -c ./common/config/.golangci-format.yml ./tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go
INFO [config_reader] Used config file common/config/.golangci-format.yml
INFO [lintersdb] Active 2 linters: [gci goimports]
INFO [loader] Go packages loading at mode 575 (compiled_files|imports|name|deps|exports_file|files|types_sizes) took 1.363456669s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 271.585µs
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner] Issues before processing: 4, after processing: 3
INFO [runner] Processors filtering stat (out/in): cgo: 4/4, exclude-rules: 4/4, source_code: 3/3, path_shortener: 3/3, severity-rules: 3/3, skip_files: 4/4, autogenerated_exclude: 4/4, exclude: 4/4, diff: 3/3, max_same_issues: 3/3, max_from_linter: 3/3, identifier_marker: 4/4, nolint: 4/4, uniq_by_line: 3/4, max_per_file_from_linter: 3/3, path_prefixer: 3/3, sort_results: 3/3, filename_unadjuster: 4/4, path_prettifier: 4/4, skip_dirs: 4/4
INFO [runner] processing took 384.582µs with stages: nolint: 109.692µs, exclude: 102.438µs, identifier_marker: 55.864µs, autogenerated_exclude: 31.959µs, skip_files: 27.162µs, path_prettifier: 20.812µs, source_code: 17.627µs, skip_dirs: 6.88µs, uniq_by_line: 4.364µs, max_from_linter: 1.663µs, cgo: 1.464µs, path_shortener: 1.248µs, max_per_file_from_linter: 1.185µs, filename_unadjuster: 791ns, max_same_issues: 419ns, diff: 347ns, exclude-rules: 248ns, severity-rules: 202ns, sort_results: 120ns, path_prefixer: 97ns
INFO [runner] linters took 335.724303ms with stages: goanalysis_metalinter: 335.283167ms
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:21: File is not `gci`-ed with -local istio.io/ (gci)
        "istio.io/istio/pkg/test/framework/resource"
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:29: File is not `gci`-ed with -local istio.io/ (gci)
        "istio.io/istio/pkg/test/framework/label"
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:24: File is not `goimports`-ed with -local istio.io/ (goimports)

INFO File cache stats: 1 entries of total size 2.9KiB
INFO Memory: 19 samples, avg is 89.6MB, max is 139.1MB
INFO Execution took 1.703670308s
```

Version: v1.30.0

Config file:

```yaml
service:
  # When updating this, also update the version stored in docker/build-tools/Dockerfile in the istio/tools repo.
  golangci-lint-version: 1.27.x # use the fixed version to not introduce new linters unexpectedly
run:
  # timeout for analysis, e.g. 30s, 5m, default is 1m
  deadline: 20m

  #     vendor$, third_party$, testdata$, examples$, Godeps$, builtin$
  skip-dirs:
  - genfiles$
  - vendor$
  skip-files:
  - ".*\\.pb\\.go"
  - ".*\\.gen\\.go"

linters:
  disable-all: true
  enable:
  - goimports
  - gci
  fast: false

linters-settings:
  goimports:
    # put imports beginning with prefix after 3rd-party packages;
    # it's a comma-separated list of prefixes
    local-prefixes: istio.io/

issues:

  # Maximum issues count per one linter. Set to 0 to disable. Default is 50.
  max-per-linter: 0

  # Maximum count of issues with the same text. Set to 0 to disable. Default is 3.
  max-same-issues: 0
```

**Comment by howardjohn** [comment]
@daixiang0

**Comment by daixiang0** [comment]
Would look at it.

**Comment by daixiang0** [comment]
@howardjohn i test with istio master:
```
root:[istio]$ golangci-lint run -v --disable-all -Egci ./tests/integration/telemetry/tracing/...
INFO [config_reader] Config search paths: [./ /root/gopkg/src/github.com/daixiang0/istio/tests/integration/telemetry/tracing /root/gopkg/src/github.com/daixiang0/istio/tests/integration/telemetry /root/gopkg/src/github.com/daixiang0/istio/tests/integration /root/gopkg/src/github.com/daixiang0/istio/tests /root/gopkg/src/github.com/daixiang0/istio /root/gopkg/src/github.com/daixiang0 /root/gopkg/src/github.com /root/gopkg/src /root/gopkg /root /]
INFO [config_reader] Used config file ../.golangci.yml
INFO [lintersdb] Active 8 linters: [gci goconst gofmt golint gosimple govet misspell whitespace]
INFO [loader] Go packages loading at mode 575 (types_sizes|deps|exports_file|files|compiled_files|imports|name) took 1m6.141092475s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 510.864µs
INFO [linters context/goanalysis] analyzers took 2m32.402400832s with top 10 stages: buildir: 2m25.615601214s, inspect: 3.59563285s, ctrlflow: 1.761896268s, printf: 1.402166408s, gci: 16.101728ms, gofmt: 2.272952ms, misspell: 2.20097ms, golint: 1.185457ms, S1038: 624.703µs, S1025: 330.393µs
INFO [runner] Issues before processing: 9, after processing: 1
INFO [runner] Processors filtering stat (out/in): exclude: 1/9, exclude-rules: 1/1, nolint: 1/1, max_per_file_from_linter: 1/1, path_prettifier: 9/9, filename_unadjuster: 9/9, skip_dirs: 9/9, identifier_marker: 9/9, uniq_by_line: 1/1, path_prefixer: 1/1, cgo: 9/9, diff: 1/1, max_same_issues: 1/1, max_from_linter: 1/1, severity-rules: 1/1, skip_files: 9/9, source_code: 1/1, path_shortener: 1/1, sort_results: 1/1, autogenerated_exclude: 9/9
INFO [runner] processing took 1.16347ms with stages: exclude: 551.649µs, identifier_marker: 230.163µs, nolint: 172.292µs, autogenerated_exclude: 74.536µs, path_prettifier: 58.763µs, skip_dirs: 39.6µs, source_code: 19.025µs, cgo: 3.553µs, uniq_by_line: 3.37µs, filename_unadjuster: 2.679µs, max_same_issues: 2.236µs, max_from_linter: 1.47µs, max_per_file_from_linter: 1.185µs, path_shortener: 934ns, diff: 429ns, skip_files: 382ns, severity-rules: 377ns, sort_results: 296ns, exclude-rules: 279ns, path_prefixer: 252ns
INFO [runner] linters took 51.352329267s with stages: goanalysis_metalinter: 51.351062325s
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:24: File is not `gci`-ed (gci)

INFO File cache stats: 4 entries of total size 13.4KiB
INFO Memory: 1086 samples, avg is 418.6MB, max is 1292.4MB
INFO Execution took 1m57.500029975s
```

**Comment by howardjohn** [comment]
@daixiang0 you are running with some custom config that does not match ours I think. See `Used config file ../.golangci.yml` from the log. Please reproduce with these steps:

checkout cfc56f29aa6091bdc85b8c7875c10d78033c366c, then run `make shell`. This gives a reproducible environment in docker.

```
$ make shell
build-tools:/work$ golangci-lint run -v --disable-all -Egci ./tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go
INFO [config_reader] Config search paths: [./ /work/tests/integration/telemetry/tracing/clienttracing /work/tests/integration/telemetry/tracing /work/tests/integration/telemetry /work/tests/integration /work/tests /work /]
INFO [lintersdb] Active 1 linters: [gci]
INFO [loader] Go packages loading at mode 575 (files|name|types_sizes|exports_file|deps|imports|compiled_files) took 1.478951509s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 224.226µs
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner] Processors filtering stat (out/in): cgo: 1/1, autogenerated_exclude: 1/1, identifier_marker: 1/1, exclude: 1/1, exclude-rules: 1/1, filename_unadjuster: 1/1, skip_dirs: 1/1, nolint: 1/1, diff: 1/1, max_same_issues: 1/1, severity-rules: 1/1, sort_results: 1/1, path_prettifier: 1/1, skip_files: 1/1, max_per_file_from_linter: 1/1, source_code: 1/1, uniq_by_line: 1/1, max_from_linter: 1/1, path_shortener: 1/1, path_prefixer: 1/1
INFO [runner] processing took 375.58µs with stages: nolint: 163.169µs, exclude: 57.655µs, autogenerated_exclude: 52.078µs, identifier_marker: 36.116µs, path_prettifier: 30.208µs, source_code: 18.878µs, skip_dirs: 8.198µs, uniq_by_line: 1.967µs, max_same_issues: 1.895µs, filename_unadjuster: 1.05µs, cgo: 998ns, max_from_linter: 892ns, path_shortener: 786ns, max_per_file_from_linter: 480ns, diff: 335ns, skip_files: 224ns, exclude-rules: 210ns, severity-rules: 200ns, sort_results: 131ns, path_prefixer: 110ns
INFO [runner] linters took 326.637969ms with stages: gci: 326.143217ms
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:25: File is not `gci`-ed (gci)

INFO File cache stats: 1 entries of total size 2.9KiB
INFO Memory: 20 samples, avg is 88.6MB, max is 139.1MB
INFO Execution took 1.853354773s
build-tools:/work$ golangci-lint run -v --disable-all -Egci ./tests/integration/telemetry/tracing/...
INFO [config_reader] Config search paths: [./ /work/tests/integration/telemetry/tracing /work/tests/integration/telemetry /work/tests/integration /work/tests /work /]
INFO [lintersdb] Active 1 linters: [gci]
INFO [loader] Go packages loading at mode 575 (deps|files|imports|name|compiled_files|exports_file|types_sizes) took 1.305011191s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 312.023µs
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner] processing took 2.952µs with stages: skip_dirs: 527ns, max_same_issues: 414ns, nolint: 213ns, autogenerated_exclude: 190ns, path_prettifier: 188ns, filename_unadjuster: 181ns, max_from_linter: 171ns, cgo: 140ns, identifier_marker: 128ns, path_prefixer: 126ns, uniq_by_line: 124ns, skip_files: 124ns, diff: 123ns, max_per_file_from_linter: 50ns, source_code: 46ns, exclude-rules: 43ns, path_shortener: 43ns, exclude: 41ns, sort_results: 40ns, severity-rules: 40ns
INFO [runner] linters took 297.561204ms with stages: gci: 297.526431ms
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 18 samples, avg is 86.9MB, max is 139.1MB
INFO Execution took 1.607461811s
build-tools:/work$
```

**Comment by SVilgelm** [comment]
@daixiang0 any updates on this?

**Comment by ldez** [comment]
The problem has been fixed in v1.34.0

**Comment by daixiang0** [comment]
Sorry that I am busy with other things, would check it later.

**Comment by daixiang0** [comment]
/reopen since not fixed yet:
```
build-tools:/work# /go/bin/golangci-lint run -v --disable-all -Egci ./tests/integration/telemetry/tracing/... --timeout 10m
INFO [config_reader] Config search paths: [./ /work/tests/integration/telemetry/tracing /work/tests/integration/telemetry /work/tests/integration /work/tests /work / /home]
INFO [lintersdb] Active 1 linters: [gci]
INFO [loader] Go packages loading at mode 575 (imports|name|types_sizes|compiled_files|deps|exports_file|files) took 2.434903061s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 441.76µs
INFO [linters context/goanalysis] analyzers took 133.364µs with top 10 stages: gci: 133.364µs
INFO [runner] processing took 7.766µs with stages: max_same_issues: 1.221µs, nolint: 937ns, skip_dirs: 809ns, cgo: 553ns, path_prettifier: 528ns, max_from_linter: 420ns, source_code: 381ns, skip_files: 364ns, diff: 347ns, uniq_by_line:343ns, identifier_marker: 286ns, autogenerated_exclude: 267ns, filename_unadjuster: 262ns, path_shortener: 251ns, exclude: 193ns, max_per_file_from_linter: 178ns, exclude-rules: 116ns, severity-rules: 112ns, path_prefixer: 102ns, sort_results: 96ns
INFO [runner] linters took 484.624887ms with stages: gci: 484.531503ms
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 31 samples, avg is 84.4MB, max is 138.6MB
INFO Execution took 2.928427607s
bash: autojump_add_to_database: command not found
build-tools:/work# /go/bin/golangci-lint run -v --disable-all -Egci ./tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go --timeout 10m
INFO [config_reader] Config search paths: [./ /work/tests/integration/telemetry/tracing/clienttracing /work/tests/integration/telemetry/tracing /work/tests/integration/telemetry /work/tests/integration /work/tests /work / /home]
INFO [lintersdb] Active 1 linters: [gci]
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|deps|exports_file|name|files|imports) took 2.498566905s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 381.299µs
INFO [linters context/goanalysis] analyzers took 251.519µs with top 10 stages: gci: 251.519µs
INFO [runner] Processors filtering stat (out/in): autogenerated_exclude: 1/1, exclude-rules: 1/1, nolint: 1/1, uniq_by_line: 1/1, max_same_issues: 1/1, path_shortener: 1/1, filename_unadjuster: 1/1, identifier_marker: 1/1, max_per_file_from_linter: 1/1, max_from_linter: 1/1, source_code: 1/1, severity-rules: 1/1, path_prefixer: 1/1, path_prettifier: 1/1, skip_files: 1/1, skip_dirs: 1/1, exclude: 1/1, diff: 1/1, sort_results: 1/1, cgo: 1/1
INFO [runner] processing took 444.74µs with stages: nolint: 267.146µs, autogenerated_exclude: 47.623µs, identifier_marker: 41.134µs, path_prettifier: 28.02µs, source_code: 25.196µs, exclude-rules: 12.075µs, skip_dirs: 9.07µs, uniq_by_line: 2.98µs, max_same_issues: 2.5µs, cgo: 1.633µs, max_from_linter: 1.341µs, path_shortener: 1.34µs, filename_unadjuster: 1.271µs, max_per_file_from_linter: 792ns, diff: 786ns, severity-rules: 571ns, exclude: 367ns, skip_files: 359ns, sort_results: 300ns, path_prefixer: 236ns
INFO [runner] linters took 506.452075ms with stages: gci: 505.906217ms
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:25: File is not `gci`-ed (gci)

INFO File cache stats: 1 entries of total size 2.9KiB
INFO Memory: 32 samples, avg is 88.1MB, max is 138.6MB
INFO Execution took 3.012588824s
bash: autojump_add_to_database: command not found
build-tools:/work# /go/bin/golangci-lint version
golangci-lint has version 1.34.0 built from 260dc80 on 2020-12-28T16:20:11Z

```

**Comment by ldez** [comment]
There is no bug with gci. All the linters will have the same behavior in this context.

The files are just excluded because of the build tag.

```
// +build integ
```

- https://github.com/istio/istio/blob/cfc56f29aa6091bdc85b8c7875c10d78033c366c/tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go#L1-L16
- https://github.com/istio/istio/blob/cfc56f29aa6091bdc85b8c7875c10d78033c366c/tests/integration/telemetry/tracing/servertracing/tracing_test.go#L1-L16

```console
$ git checkout cfc56f29aa6091bdc85b8c7875c10d78033c366c
Note: switching to 'cfc56f29aa6091bdc85b8c7875c10d78033c366c'.
...
$ golangci-lint run --no-config --disable-all  -Egci ./tests/integration/telemetry/tracing/...

$ golangci-lint run --no-config --disable-all --build-tags integ -Egci ./tests/integration/telemetry/tracing/...
tests/integration/telemetry/tracing/clienttracing/client_tracing_test.go:25: File is not `gci`-ed (gci)
```

The build tags must be defined via `--build-tags`, ex: ` --build-tags integ`

This behavior comes from the `analysis` package (`golang.org/x/tools`)

So everything works as expected.
